% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/MVP.MLM.R
\name{MVP.MLM}
\alias{MVP.MLM}
\title{To perform GWAS with GLM and MLM model and get the P value of SNPs}
\usage{
MVP.MLM(phe, geno, K = NULL, CV = NULL, REML = NULL,
  priority = "speed", cpu = 2, bar = TRUE, vc.method = "EMMA",
  maxLine = 1000, file.output = TRUE, memo = "MVP")
}
\arguments{
\item{phe}{phenotype, n * 2 matrix}

\item{geno}{genotype, m * n, m is marker size, n is population size}

\item{K}{Kinship, Covariance matrix(n * n) for random effects; must be positive semi-definite}

\item{CV}{covariates}

\item{REML}{a list that contains ve and vg}

\item{priority}{speed or memory}

\item{cpu}{number of cpus used for parallel computation}

\item{bar}{whether to show the progress bar}

\item{vc.method}{the methods for estimating variance component("emma" or "gemma")}

\item{maxLine}{when the priority is 'memory', users can change this parameter to limit the memory}

\item{file.output}{whether to output files or not}

\item{memo}{a marker on output file name}
}
\value{
results: a m * 2 matrix, the first column is the SNP effect, the second column is the P values
}
\description{
Build date: Aug 30, 2016
Last update: Aug 30, 2016
}
\examples{
phePath <- system.file("extdata", "mvp.phe", package = "rMVP")
phenotype <- read.table(phePath, header=TRUE)
print(dim(phenotype))
genoPath <- system.file("extdata", "mvp.geno.desc", package = "rMVP")
genotype <- attach.big.matrix(genoPath)
print(dim(genotype))
K <- MVP.K.VanRaden(genotype)
mlm <- MVP.MLM(phe=phenotype, geno=genotype, K=K)
str(mlm)
}
\author{
Lilin Yin and Xiaolei Liu
}
